pkg/lib/cockpit-upload-helper.ts

```diff
@@ -0,0 +1,133 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import cockpit from 'cockpit';
+
+function debug(...args: unknown[]) {
+    if (window.debugging == 'all' || window.debugging?.includes('upload'))
+        console.debug('debug[upload]', ...args);
+}
+
+class UploadError extends Error {
+    name = 'UploadError';
+    constructor(problem: string, message: cockpit.JsonObject) {
+        // XXX message.message is untranslated.  Should we just always use cockpit.message(problem)?
+        super(typeof message.message === 'string' ? message.message : cockpit.message(problem));
+    }
+}
+
+/* XXX: these values are ~a decade old and were more or less copied from ssh
+ * (where they were last changed in 2007).  We might test alternatives...
+ */
+const BLOCK_SIZE = 16 << 10; // 16kiB
+const FLOW_WINDOW = 2 << 20; // 2MiB
+
+export async function upload(
+    destination: string,
+    stream: ReadableStream,
+    progress?: (bytes_sent: number) => void,
+    signal?: AbortSignal,
+    options?: cockpit.JsonObject
+) {
+    /* While the upload is running, the function is essentially always waiting
+     * inside of a Promise resolved by notify().  Anything that might cause
+     * forward progress (ack messages, close messages, cancellation) will call
+     * notify().  It's up to the main loop body to re-evaluate and respond.
+     */
+    let outstanding = 0; // for flow control
+    let delivered = 0; // for progress reporting
+    let close_message = null as cockpit.JsonObject | null;
```

Comment:
I had to read this twice -- this is supposed to be `null as (cockpit.JsonObject | null)`, not `(null as cockpit.JsonObject) | null`, right?

Reply:
Yes.  And the reason for requiring this is more subtle than you probably realize.  Because `close_message` gets assigned from inside of a callback, tsc has no way of knowing that this variable will ever acquire a value other than `null`, which causes it to assume that the `while()` will never exit.  Control flow analysis across functions is not a strong point for typescript, it seems... :(

Shall I add parens?

Reply:
It's mostly me being new to type annotations, and `|` vs. `as` precedence isn't very intuitive. So I leave that up to you, feel free to just resolve.

Reply:
I added parens.  I'll leave out the comment about why it was required ;)